R/stat-sina.r

```diff
@@ -0,0 +1,196 @@
+#' @param adjust adjusts the bandwidth of the density kernel when
+#' \code{method == "density"} (see \code{\link[stats]{density}}) or the spread
+#' of the samples within the same group along the x-axis.
+#' @export
+stat_sina <-function(mapping = NULL, data = NULL,
+                     geom = "point", position = "identity",
+                     ...,
+                     binwidth = 0.02,
+                     scale = TRUE,
+                     neighbour_limit = 1,
+                     method = c("density", "neighbourhood"),
+                     adjust = 1,
+                     show.legend = NA,
+                     inherit.aes = TRUE) {
+  layer(
+    data = data,
+    mapping = mapping,
+    stat = StatSina,
+    geom = geom,
+    position = position,
+    show.legend = show.legend,
+    inherit.aes = inherit.aes,
+    params = list(
+      binwidth = binwidth,
+      scale = scale,
+      neighbour_limit = neighbour_limit,
+      method = match.arg(method),
+      adjust = adjust,
+      ...
+      )
+  )
+}
+
+
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
+#' @export
+StatSina <- ggproto("StatSina", Stat,
+
+  required_aes = c("x", "y"),
+
+  setup_data = function(data, params) {
+
+    if (is.double(data$x) && !has_groups(data) && any(data$x != data$x[1L])) {
+      stop(
+        "Continuous x aesthetic -- did you forget aes(group=...)?",
+        call. = FALSE)
+    }
+
+    data
+  },
+
+  compute_panel = function(data, scales, binwidth = 0.02, scale = TRUE,
+                           neighbour_limit = 1, method = "d", adjust = 1) {
+
+    ###Initialise variables
+
+    #neighbour counts
+    neighbours <- list()
+    max_neighbours <- 0
+
+    #method == "density"
+    if (method == "density") {
+      max_density <- 0
+      densities <- list()
+    } else {
+      #scale adjust value
+      adjust <- adjust / 5
+    }
+
+    #bin the y-axis
+    bins <- bin_y(data$y, binwidth)
+
+    # ------------------------------------------------------------------------ #
+
+    #Compute per class density and per bin sample count
+    for (j in levels(factor(data$group))) {
+
+      #extract samples per group and store them in a data.frame
+      keep <- data$group == j
+
+      if (sum(keep) < 2)
+        next
+
+      #per bin sample count
+      neighbours[[j]] <- table(findInterval(data$y[keep], bins))
+
+      #find the densiest neighbourhood in the current group and compare it
+      #with the global max
+      tmp_max_neighbours <- max(neighbours[[j]])
+
+      if (tmp_max_neighbours > max_neighbours)
+        max_neighbours <- tmp_max_neighbours
+
+      if (method == "density") {
+
+        #per bin sample density
+        densities[[j]] <- stats::density(data$y[keep], adjust = adjust)
+
+        #find the highest density value
+        tmp_max_density <- max(densities[[j]]$y)
+
+        if (tmp_max_density > max_density)
+          max_density <- tmp_max_density
+      }
```

Comment:
This (103-118) chunk has some dead code? `max_neighbours` and `max_density` are not used.